src/libaktualizr/uptane/tuf.h

```diff
@@ -153,6 +153,7 @@ class BaseMeta {
   TimeStamp expiry() const { return expiry_; }
   bool isExpired(const TimeStamp &now) const { return expiry_.IsExpiredAt(now); }
   Json::Value original() const { return original_object_; }
+  std::string signature() const { return original_object_["signatures"][0]["sig"].asString(); }
```

Comment:
This makes me just a bit uncomfortable, because there's a lot of assumptions packed in this one line, and it can throw really easily. At least a `isMember("signatures")` check would help.

Reply:
> then we make an assumption that a signature verification happens before this call

I think a decent compromise for now is just to state that assumption as a comment.

> Should I just throw exception if the check of "signatures"/0/"sig" fails? I think so since if they are missing then metadata is invalid by definition.

That wouldn't hurt, either.

> The only concern I have is that timestamp can have more than one signatures and the first signature is the same for both versions while the rest different, but I don't think that's possible as long as the content/subject is the same.

You could also have multiple signatures with different algorithms. We experimented with that once upon a time, but it never went anywhere. I think the single-signature assumption should also be stated in a comment.

Reply:
> The only concern I have is that timestamp can have more than one signatures and the first signature is the same for both versions while the rest different, but I don't think that's possible as long as the content/subject is the same.

We do have 2 signatures for targets for sure in the case of production targets at Foundries.io.  So it can be easily imagined that somewhere in the different place or time there might be two signature for the timestamp as well. At one point I considered doing that, so might return back to this topic in the future.

That said, an assumption that there is just one signature seems like a timebomb.  From the other perspective, if one signature is equal there is no reason why would the second one be different.

Reply:
@mike-sul I've checked in our production targets implementation that we use the consistent order of signatures i.e. the signature for the same key always goes at the same position in the list.  So this patch should work as is.

But we should definitely update this later on with a more consistent solution.
IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

Reply:
> IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

If signature order is different in the same versions of different metadata then we are fine because  stored_sign != update_sign and we store the new version of metadata.

If signature order is different in the same versions of the same metadata then we are fine too because  stored_sign != update_sign and we just store the same metadata again.

Reply:
> IIRC we've had a long discussion a year ago about why HAVING consistent order for signatures is a good thing while RELYING onto this consistent order is a bad thing.

Fully agree.

> If signature order is different in the same versions of the same metadata then we are fine too because stored_sign != update_sign and we just store the same metadata again.

That is a fair point. Not super efficient, but probably fine. My thought is that there probably isn't a lot of need to worry about this now, but if, down the line, there are more complexities around the signatures, this would be worth revisiting.